### PR TITLE
MUIC-14:[Android] File Attachments Upload

### PR DESCRIPTION
**Pull request by deniss-mc**

### CODE REVIEW

File: widgetssdk/src/main/res/layout/chat_attachment_item_file.xml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                app:layout_constraintHorizontal_bias="0"
+                app:layout_constraintStart_toEndOf="@id/type_indicator_view"
+                app:layout_constraintTop_toBottomOf="@id/status_indicator"
+                app:layout_constraintWidth_percent="0.5"
+                app:trackColor="@color/glia_system_agent_bubble_color"
+                app:trackCornerRadius="4dp" />
+
+        </androidx.constraintlayout.widget.ConstraintLayout>
+    </com.google.android.material.card.MaterialCardView>
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/res/layout/chat_attachment_item_file.xml:100]: Too many nested views might cause performance issues

File: widgetssdk/src/main/res/layout/chat_attachment_select_item_menu.xml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                    app:layout_constraintBottom_toBottomOf="parent"
+                    app:layout_constraintEnd_toEndOf="parent"
+                    app:layout_constraintTop_toTopOf="parent" />
+
+            </androidx.constraintlayout.widget.ConstraintLayout>
+
+        </LinearLayout>
+    </androidx.cardview.widget.CardView>
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/res/layout/chat_attachment_select_item_menu.xml:146]: nested views

File: widgetssdk/src/main/java/com/glia/widgets/fileupload/FileAttachmentRepository.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.List;
+import java.util.Observable;
+import java.util.Observer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FileAttachmentRepository {
+    private static class ObservableFileAttachmentList extends Observable {
+        public List<FileAttachment> fileAttachments = new ArrayList<>();
+        public void notifyUpdate(List<FileAttachment> newObject) {
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/fileupload/FileAttachmentRepository.java:20]: a line of space maybe

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatVisitorMessageAttachmentAdapter.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            long byteSize = attachmentFile.getSize();
+
+            extensionContainerView.setCardBackgroundColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_brand_primary_color));
+            titleText.setText(String.format("%s • %s", name, Formatter.formatFileSize(itemView.getContext(), byteSize)));
+            String extension = Utils.getExtensionByStringHandling(name).orElse("");
+            extensionTypeText.setText(extension);
+        }
+    }
+
+    private interface ItemCallback {
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatVisitorMessageAttachmentAdapter.java:156]: is this needed?

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatVisitorMessageAttachmentAdapter.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                }
+            }
+        }
+    }
+
+    static class FileViewHolder extends RecyclerView.ViewHolder {
+        public FileViewHolder(@NonNull View itemView) {
+            super(itemView);
+        }
+
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatVisitorMessageAttachmentAdapter.java:127]: extra line

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
+        holder.onBind(getItem(position), callback);
+    }
+
+    static class ViewHolder extends RecyclerView.ViewHolder {
+        public ViewHolder(@NonNull View itemView) {
+            super(itemView);
+        }
+
+        public void onBind(FileAttachment attachment, ItemCallback callback) {
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java:71]: some parts can be extracted as functions

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+                    break;
+                case ERROR_FILE_TOO_LARGE:
+                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
+                    titleText.setText(R.string.chat_attachment_upload_error_file_size_over_limit);
+                    break;
+                case ERROR_ENGAGEMENT_MISSING:
+                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
+                    titleText.setText(R.string.chat_attachment_upload_error_engagement_missing);
+                    break;
+                case ERROR_UNKNOWN:
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java:172]: case body missing

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private void setTitleText(TextView titleText, String fileName, long byteSize, FileAttachment.Status status) {
+            switch (status) {
+                case UPLOADING:
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java:142]: case bodies missing

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            }
+        }
+
+        private String getStatusIndicatorText(Context context, FileAttachment.Status status) {
+            switch (status) {
+                case SECURITY_SCAN:
+                    return context.getString(R.string.chat_attachment_upload_checking_file);
+                case READY_TO_SEND:
+                    return context.getString(R.string.chat_attachment_upload_ready_to_send);
+                case ERROR_NETWORK_TIMEOUT:
```
Comment (mustafa-ogun-ozturk) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java:218]: case bodies missing

**Review by mustafa-ogun-ozturk — APPROVED** [review]

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatView.java [hunk 10]
```diff
@@ -756,6 +845,27 @@ public void afterTextChanged(Editable editable) {
         });
     }
 
+    private void dispatchImageCapture() {
+        Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+        File photoFile = null;
+        try {
+            photoFile = Utils.createTempPhotoFile(getContext());
+        } catch (IOException exception) {
+            exception.printStackTrace();
```
[the remark this answers is not in the document]
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatView.java:854]: No message available for "Storage full" or any other exceptions atm. we might deal with that later

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatVisitorMessageAttachmentAdapter.java (hunk 4 above)
Comment (deniss-mc) [reply widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatVisitorMessageAttachmentAdapter.java:156]: Will be needed for file download callback

**Pull request merged**